doc/GUIDE.md

```diff
@@ -0,0 +1,264 @@
+
+rx -r pcre -l ast -d declined.txt 'x*?'
+# Unsupported character in declined.txt
+```
+
+If unsupported constructs exist, libfsm reports the failing location.
+
+### 3. Generate Code
+
+```bash
+re -p -r pcre -l rust -k str '^item-[A-Z]{3}\z' > item_detector.rs
+```
+
+### 4. Multiple Patterns
+
+```bash
+# re - patterns from command line:
+re -p -r pcre -l go -k str '^x?a b+c$' '^x*def?$' '^x$'
+
+# rx - patterns from file:
+rx -p -r pcre -l vmc -k str -d skipped.txt patterns.txt > detectors.c
+```
+
+Both tools:
+* Combine all patterns into one function (like using `|` to join them)
+* Return `(bool, int)` - match status and pattern ID
+* Pattern ID is argument position for `re`, line number for `rx`
+* When encountering unsupported patterns: `rx` skips them to `-d` file and generates code with working patterns; `re` fails completely
+
+---
+
+### Flag Reference
+| Flag | Purpose                      | Common Options                             | Notes                                                            |
+| ---- | ---------------------------- | ------------------------------------------ | ---------------------------------------------------------------- |
+| `-r` | Regex dialect                | `pcre`, `literal`, `glob`, `native`, `sql` | `pcre` supports the widest set of features                       |
+| `-l` | Output language for printing | `go`, `rust`, `vmc`, `llvm`, `wasm`, `dot` | Use `vmc` for `C` code. Pipe `dot` into `idot` for visualization |
+| `-k` | Generated function I/O API   | `str`, `getc`, `pair`                      | `str` takes string, `pair` takes byte array, `getc` uses callback for streaming |
+| `-p` | Print mode                   | *(no value)*                               | Abbrv. of `-l fsm`. Print the constructed fsm, rather than executing it.        |
+| `-d` | Declined                     | filename                                   | Only applies to `rx` (batch mode)                                |
+
+This is not exhausted list. For full flag details, see [include/fsm/options.h](../include/fsm/options.h) and the [man pages](../man).
+The man pages can be built by running `bmake doc`, then view with `build/man/re.1/re.1`.
+
+---
+
+## Writing Effective libfsm Patterns
+
+### 1. Replace Broad Wildcards
+
+Avoid `.*` and `.+` when possible. Wildcards match “anything,” which is often imprecise and forces libfsm to build a large DFA.
+
+For example, a double-quoted string should not use `".*?"` because the content cannot contain an unescaped quote.
+Instead, restrict it to the actual valid characters `"[^"\r\n]*"`, which matches only what is allowd and will keep the DFA more compact.
+
+Use negated character classes:
+
+| Avoid      | Better         |
+| ---------- | -------------- |
+| `<.*>`     | `<[^>]*>`      |
+| `\((.*)\)` | `\([^)]*\)`    |
+| `price=.+` | `price=[0-9]+` |
+| `var\s.+=` | `var\s[^=]+=`  |
+
+> This is often the cause of an “explosion” in the size of the generated FSM.
```

Comment:
for technical writing, it's almost always a good idea to say explicitly what you're talking about instead of "This". in this case, that's more precisely:
> The overlap between `.*` and strings that follow is often the cause of an ...